app/code/core/Mage/Core/Model/Encryption.php

```diff
@@ -153,53 +148,61 @@ public function validateHashByVersion(#[\SensitiveParameter] $password, $hash, $
         return hash_equals($this->hash($salt . $password, $version), $hash);
     }
 
-    /**
-     * Instantiate crypt model
-     *
-     * @param string $key
-     * @return Varien_Crypt_Mcrypt
-     */
-    protected function _getCrypt($key = null)
+    public function encrypt(#[\SensitiveParameter] string $data): string
     {
-        if (!$this->_crypt) {
-            if ($key === null) {
-                $key = (string) Mage::getConfig()->getNode('global/crypt/key');
-            }
-            $this->_crypt = Varien_Crypt::factory()->init($key);
-        }
-        return $this->_crypt;
+        $key = Mage::getEncryptionKeyAsBinary();
+        $nonce = random_bytes(SODIUM_CRYPTO_SECRETBOX_NONCEBYTES);
+        $ciphertext = sodium_crypto_secretbox($data, $nonce, $key);
+        $encrypted = sodium_bin2base64($nonce . $ciphertext, SODIUM_BASE64_VARIANT_ORIGINAL);
+
+        // Clean sensitive data from memory
+        sodium_memzero($data);
+        sodium_memzero($key);
+        sodium_memzero($nonce);
+        sodium_memzero($ciphertext);
+
+        return $encrypted;
     }
 
-    /**
-     * Encrypt a string
-     *
-     * @param string $data
-     * @return string
-     */
-    public function encrypt($data)
+    public function decrypt(#[\SensitiveParameter] string $data): string
     {
```

Comment:
```suggestion
    public function decrypt(#[\SensitiveParameter] ?string $data): string
    {
        if (empty($data)) {
            // TODO: also add exception
            return '';
        }
```

I restored by database from a backup from before I set up 2fa, then went to the My Account page and got this error:

`Fatal error:  Uncaught TypeError: Mage_Core_Model_Encryption::decrypt(): Argument #1 ($data) must be of type string, null given, called in /home/www-data/work/maho/git/app/code/core/Mage/Admin/Model/User.php on line 979 and defined in /home/www-data/work/maho/git/app/code/core/Mage/Core/Model/Encryption.php:167`

So before we must have allowed to decrypt a null value.

Reply:
I did it this way if that's ok with you

<img width="584" alt="Screenshot 2025-03-31 alle 18 48 04" src="https://github.com/user-attachments/assets/4384aa1b-c845-42d1-a9b5-95300c28d7cb" />

cause '0' also matches empty so I think a strict comparison is better in this case.

also, I don't know, I kinda wouldn't log an exception in this case, we can be tolerant and allow people to decrypt empty strings with no errors IMO

Reply:
Yup, totally good.